test/integration_tests/integrationtest_scanner_api.cpp

```diff
@@ -165,12 +172,57 @@ TEST(ScannerAPITests, testStartFunctionality)
   EXPECT_EQ(start_future.wait_for(DEFAULT_TIMEOUT), std::future_status::ready) << "Scanner::start() not finished";
 }
 
+TEST(ScannerAPITests, shouldThrowWhenStartIsCalledTwice)
+{
+  NiceMock<ScannerMock> scanner_mock;
+  const ScannerConfiguration config{ createScannerConfig() };
+  UserCallbacks cb;
+  ScannerV2 scanner(config, std::bind(&UserCallbacks::LaserScanCallback, &cb, std::placeholders::_1));
+
+  Barrier start_req_received_barrier;
+  ON_CALL(scanner_mock, receiveControlMsg(_, StartRequest(config, DEFAULT_SEQ_NUMBER).serialize()))
+      .WillByDefault(OpenBarrier(&start_req_received_barrier));
+
+  scanner_mock.startListeningForControlMsg();
+  scanner.start();
+  EXPECT_THROW(scanner.start();, std::runtime_error);
+}
+
+TEST(ScannerAPITests, shouldNotFailInCaseOfUnexpectedMonitoringFrameDuringStart)
+{
+  StrictMock<ScannerMock> scanner_mock;
+  const ScannerConfiguration config{ createScannerConfig() };
+  UserCallbacks cb;
+  ScannerV2 scanner(config, std::bind(&UserCallbacks::LaserScanCallback, &cb, std::placeholders::_1));
+
+  Barrier start_req_received_barrier;
+  EXPECT_CALL(scanner_mock, receiveControlMsg(_, StartRequest(config, DEFAULT_SEQ_NUMBER).serialize()))
+      .WillOnce(OpenBarrier(&start_req_received_barrier));
+  EXPECT_CALL(cb, LaserScanCallback(_)).Times(0);
+
+  scanner_mock.startListeningForControlMsg();
+  const auto start_future = scanner.start();
+
+  EXPECT_TRUE(start_req_received_barrier.waitTillRelease(DEFAULT_TIMEOUT)) << "Start request not received";
+  EXPECT_EQ(start_future.wait_for(WAIT_TIMEOUT), std::future_status::timeout) << "Scanner::start() finished too early";
+
+  std::cout << "ScannerAPITests: Send unexpected monitoring frame..." << std::endl;
+  std::vector<double> measures = { 1, 2, 3, 4, 5 };
+  MonitoringFrameMsg monitoring_frame(TenthOfDegree(0), TenthOfDegree(1), 0, measures);
+  scanner_mock.sendMonitoringFrame(monitoring_frame);
+  EXPECT_EQ(start_future.wait_for(WAIT_TIMEOUT), std::future_status::timeout) << "Scanner::start() finished too early";
+
+  std::cout << "ScannerAPITests: Send start reply..." << std::endl;
+  scanner_mock.sendStartReply();
+  EXPECT_EQ(start_future.wait_for(DEFAULT_TIMEOUT), std::future_status::ready) << "Scanner::start() not finished";
+}
```

Comment:
If I understand that right this test tests:
- No laser scan callback is called on early monitoring frame
- barrier does open on start request.
- start future is set valid right after start
- start future is set valid by sending monitoring frame
- start future is valid after sending start reply.

The name states does not fail, which can mean a lot, but is all that necessary to test if an early monitoring frame does not change the start behavior?
Maybe adding one or two more tests would split this into easier understandable chunks

Reply:
>[...] but is all that necessary to test if an early monitoring frame does not change the start behavior?

I removed or changed some of the checks. However, some checks simply are necessary because the test is dealing with more than one thread. To get the timing right, one simply needs to check certain conditions.

And yes a monitoring frame should not change the start behavior and this is exactly what this test tries to check. I'm not sure how I should reduce the test complexity. The scenario is a little bit complicated to reproduce but I don't see how I should simplify the test.